packages/core/src/loop/workflows/agentic-execution/tool-call-step.ts

```diff
@@ -107,6 +117,94 @@ export function createToolCallStep<
+
+        if (lastAssistantMessage) {
+          const content = lastAssistantMessage.content;
+          if (!content) return;
+          // Add metadata to indicate this tool call is pending approval
+          const metadata =
+            typeof lastAssistantMessage.content.metadata === 'object' && lastAssistantMessage.content.metadata !== null
+              ? (lastAssistantMessage.content.metadata as Record<string, any>)
+              : {};
+          metadata.suspendedTools = metadata.suspendedTools || {};
+          metadata.suspendedTools[toolName] = {
+            toolCallId,
+            toolName,
+            args,
+            type: 'suspension',
+            runId, // Store the runId so we can resume after page refresh
+            suspendPayload,
+            resumeSchema,
+          };
+          lastAssistantMessage.content.metadata = metadata;
+        }
+      };
+
+      // Helper function to remove tool suspension metadata after resume
+      const removeToolSuspensionMetadata = async (toolName: string) => {
+        const { saveQueueManager, memoryConfig, threadId } = _internal || {};
+
+        if (!saveQueueManager || !threadId) {
+          return;
+        }
+
+        const getMetadata = (message: MastraDBMessage) => {
+          const content = message.content;
+          if (!content) return undefined;
+          const metadata =
+            typeof content.metadata === 'object' && content.metadata !== null
+              ? (content.metadata as Record<string, any>)
+              : undefined;
+          return metadata;
+        };
+
+        // Find and update the assistant message to remove approval metadata
+        // At this point, messages have been persisted, so we look in all messages
+        const allMessages = messageList.get.all.db();
+        const lastAssistantMessage = [...allMessages].reverse().find(msg => {
+          const metadata = getMetadata(msg);
+          const suspendedTools = metadata?.suspendedTools as Record<string, any> | undefined;
+          return !!suspendedTools?.[toolName];
+        });
+
+        if (lastAssistantMessage) {
+          const metadata = getMetadata(lastAssistantMessage);
+          const suspendedTools = metadata?.suspendedTools as Record<string, any> | undefined;
+
+          if (suspendedTools && typeof suspendedTools === 'object') {
+            delete suspendedTools[toolName];
+
+            // If no more pending suspensions, remove the whole object
+            if (metadata && Object.keys(suspendedTools).length === 0) {
+              delete metadata.suspendedTools;
+            }
+
+            // Flush to persist the metadata removal
+            try {
+              await saveQueueManager.flushMessages(messageList, threadId, memoryConfig);
+            } catch (error) {
+              console.error('Error removing tool suspension metadata:', error);
+            }
+          }
+        }
+      };
```

Comment:
I wonder if we should merge these and the approval metadata helpers together. They seem very similar outside of a couple of code differences.